### PR TITLE
server : parallel decoding and multimodal (cont)

### LINKED ISSUE #3693: will `./server` support `parallel` decoding as well?
# Expected Behavior

The `parallel` examples look promising. I'm wondering if `./server` will also support an `-np` argument to process requests in parallel. This way, the user can send `np` prompts at a time.

# Current Behavior

Currently, `./server` processes requests sequentially.

# Environment and Context

macOS Sonoma, M1 Pro chip

### PR DESCRIPTION
**Pull request by ggerganov**
Continuation of https://github.com/ggerganov/llama.cpp/pull/3589

Commands to test:

```bash
# parallel decoding from multiple clients (`-np 4`) with continuous batching (`-cb`)
./server -m ${model} --ctx_size 2048 -t 8 -ngl 99 -np 4 -cb

# vision support, upload image in the browser and ask questions
./server -m ./models/llava-7b-v1.5/ggml-model-f16.gguf --mmproj ./models/llava-7b-v1.5/mmproj-model-f16.gguf --ctx_size 2048 -t 8 -ngl 99
```

### CODE REVIEW

**Comment by FSSRepo** [comment]
In the server's frontend (server ui), it's missing the display of the image and its properties (resolution and format) so that when users encounter an error, they can know the image properties.

**Comment by FSSRepo** [comment]
stb-image just support .png and .jpg images format

**Comment by FSSRepo** [comment]
> In the server's frontend (server ui), it's missing the display of the image and its properties (resolution and format) so that when users encounter an error, they can know the image properties.

stb-image just support .png and .jpg images format.

I don't know the maximum effective resolution that clip model support for best quality result. @monatis perhaps do you know something about it?

**Comment by monatis** [comment]
@FSSRepo Technically there's no upper limit but aspect ratios close to 1.0 are recommended because we first pad images to a square then resize them to 336 x 336. If it diverges from a square aspect ratio considerably then a larger padding is applied, and the larger the padding, the bigger the information loss.

**Comment by ggerganov** [comment]
There are many data races and race conditions in the proposed implementation. The http handlers are freely accessing and mutating the `llama_server_context` state without synchronization.

Depending on the amount of work required to fix this, this might become a show stopper - i.e. prevent from merging these PR. Will continue looking later today.

**Comment by FSSRepo** [comment]
> Depending on the amount of work required to fix this, this might become a show stopper


I can't think of any other way to do it, and using mutex and blocking conditionals would only cause serious performance losses because the processing thread `updateSlot() loop` could be undesirably blocked.

**Comment by ggerganov** [comment]
The way to fix this is to introduce 2 queues:

- In the first queue, the HTTP requests push tasks with certain input (could be simply the entire json from the client).
- In the second queue, the `update_slots()` loop pushes tasks with the generated results from the execution

The queues are thread-safe so that pushing and popping is safe. I.e. the only mutexes and condition variables used in `server.cpp` have to exist inside the queue implementation behind it's API. No locking and waiting in random places of the code.

Each request handler now looks something like this:

```c
req_id = generate_req_id();
llama.add_task(req_id, json); // pushes data in to the input queue tagging it with a unique req_id
while (true) {
    data = llama.get_result(req_id); // this blocks on wait condition waiting for the "backend" to push data for req_id
    if (data.has_finished) break;
    res.set_content(data.json, ...);
}
```

In `update_slots()`, on each iteration we loop through the input queue and process the json data from the clients in a similar way as we do currently in the HTTP handlers. But now, we have moved this in the "backend" and processing is synchronized on the main server thread.

After computing stuff, `update_slots()` pushes data back, tagging it with the respective `req_id` so that the HTTP handlers know which data is relevant for them.

These are just rough ideas how to solve the synchronization problems. It's not a complete description and could be missing some details.

Without implementing this or a similar solution, we cannot merge the PR. Unfortunately, it looks like quite some extra work would be needed to achieve this. Also, have to make sure that the thread sanitizer does not produce any errors.

**Comment by monatis** [comment]
> The way to fix this is to introduce 2 queues:

Love this idea. This is highly scalable and flexible. I remembered [Adrian Rosebrock's 2018 tutorial](https://pyimagesearch.com/2018/01/29/scalable-keras-deep-learning-rest-api/), where he implements this architecture with Redis as the queue. It also has another advantage that might be very useful optimizing batch-processing. He defines a server sleep time (say 0.25 seconds in the tutorial) to get tasks from the queue. This can be adapted based on the server load to allow accumulation of batch_size tasks, hence better utilizing the accelerator.

**Comment by FSSRepo** [comment]
> The way to fix this is to introduce 2 queues:

In c++ is there some concurrent vector or map? I will try it

**Comment by monatis** [comment]
[This stackoverflow answer](https://stackoverflow.com/a/16075550) might be helpful.

**Comment by FSSRepo** [comment]
I consider that `llama_sampling_init` should be more flexible, just `const struct gpt_params & params` is restrictive to this implementation.

The params should be `const struct llama_sampling_params &sparams, const std::string &grammar, int n_ctx` for more control when is needed.

**Comment by ggerganov** [comment]
> I consider that llama_sampling_init should be more flexible, just const struct gpt_params & params is restrictive to this implementation.

Yes, I'll open a PR later to update this

**Comment by FSSRepo** [comment]
> [This stackoverflow answer](https://stackoverflow.com/a/16075550) might be helpful.

The example of safe queue when the queue is empty that blocks the execution until is added other element.

**Comment by FSSRepo** [comment]
I did it in the best way I could, there should be no race condition at any point, as the queues are synchronized.

**Comment by ggerganov** [comment]
Ok, this is looking better - we can work with it and improve later from `master`.

I will first merge #3696 and then will help to adapt this branch to the changes

**Comment by ggerganov** [comment]
In general, all places where you have used `sleep` can be replaced with waiting on `std::condition_variable`. If you are not familiar how to do it - no worries. We'll fix it in the future, and the current solutions is good enough

**Comment by monatis** [comment]
The UI works only with png images and not with jpeg. And there's no error shown to user when it fails to load an image.

**Comment by jxy** [comment]
Does Llava allow only one image at the start of a conversation? Can we also include multiple images in subsequent turns?

**Comment by FSSRepo** [comment]
> The UI works only with png images and not with jpeg. 

Just png and jpg

**Comment by monatis** [comment]
didn't work with jpg images in my case, and only a terminal output saying "failed to load image at id= 10". Tried with several jpg images.

**Comment by ggerganov** [comment]
Doesn't work with `.jpg` for me as well. Only `.png`.

File: examples/server/public/index.html
Comment (Green-Sky) [inline examples/server/public/index.html:492]: png
any way to make it image type agnostic? like `data:image/*;` ?

Comment (FSSRepo) [reply examples/server/public/index.html:492]: It can be 

```javascript
+ Image_data.substring(image_data.indexOf("base64,") + 7)
```

**Comment by FSSRepo** [comment]
> Doesn't work with `.jpg` for me as well. Only `.png`.

I can fix this faster, but I don't know if you are working in some change of this PR

**Comment by ggerganov** [comment]
I'll get back to this on Mondey

**Comment by monatis** [comment]
Thanks, now it also works with jpg images correctly. After fixing conflicts I think it's good to merge.

**Comment by ggerganov** [comment]
Think I can take a look today - any pending changes here that I should wait before merging `master` and cleaning up?

**Review by monatis — APPROVED** [review]
lgtm

**Comment by ggerganov** [comment]
Does this branch crash for anyone else in Debug upon the first request send from the browser?
It crashes for me on macOS - only in Debug (Bus error: 10). Backtrace does not make sense - it crashes upon call to `ggml_new_graph()`

**Comment by FSSRepo** [comment]
> Does this branch crash for anyone else in Debug upon the first request send from the browser?
> It crashes for me on macOS - only in Debug (Bus error: 10). Backtrace does not make sense - it crashes upon call to `ggml_new_graph()`

How to test it debug mode in Windows?

```bash
cmake --build . --config Debug
```

Is it right?

File: examples/server/public/index.html
Comment (monatis) [inline examples/server/public/index.html:227]: Saving image data with other params to `localStorage` (and loaind it from there) may be problematic when the user re-opens their browser. It still send it to the server, but not shown on the UI in this case. What about saving it separately to `sessionStorage`, or not saving at all?

Comment (ggerganov) [reply examples/server/public/index.html:227]: Sure, I'm not sure how to do this though. Feel free to update as necessary

Comment (monatis) [reply examples/server/public/index.html:231]: Fixed.

**Comment by ggerganov** [comment]
It seems that once a client starts using one of the slots, the slot never becomes released.
Any ideas how to fix this?

**Comment by FSSRepo** [comment]
> It seems that once a client starts using one of the slots, the slot never becomes released.
> Any ideas how to fix this?

I will review it

**Comment by ggerganov** [comment]
After we fix this, we can merge

**Comment by FSSRepo** [comment]
> After we fix this, we can merge

I've encountered an error where, upon canceling a completion and sending another message, the server crashes.

**Comment by ggerganov** [comment]
> This line crashes in commit [dd1af2e](https://github.com/ggerganov/llama.cpp/pull/3677/commits/dd1af2ed35dfcf1e6842a57c4e69478e93622d89):

Does it still crash after latest commit: [00ae55b](https://github.com/ggerganov/llama.cpp/pull/3677/commits/00ae55b3883450f7cce89fd6e24109cf130fc269)

**Comment by FSSRepo** [comment]
> It seems that once a client starts using one of the slots, the slot never becomes released. Any ideas how to fix this?

When the `cache_prompt` parameter is set to `true`, the slot is not released; instead, it is only suspended without clearing the previous state (processed prompt and previously generated tokens). That's why the client needs to send the assigned slot ID when making a request. The only issue with this is that when the client closes the web ui, the slot remains suspended indefinitely.

**Comment by FSSRepo** [comment]
> Does it still crash after latest commit: [00ae55b](https://github.com/ggerganov/llama.cpp/pull/3677/commits/00ae55b3883450f7cce89fd6e24109cf130fc269)

Fixed

File: examples/server/server.cpp
Comment (monatis) [inline examples/server/server.cpp:1487]: Is the issue with the slot not released due to this line? With `cache_prompt` it sleeps instead of being released.

**Comment by ggerganov** [comment]
> To avoid this, my idea is to add a timer to the slot and periodically send the signal to keep the slot suspended (while the user is using the web UI). If the slot doesn't receive the signal after the timer's duration, it will be automatically released.

Can't we release the slot after every request?
The logic is that a user provides the `slot_id` they've used last time and if it is released, we continue as we currently do. If it is not free, we search for a new slot. If a new request comes in with unassigned slot, we assign the one that has never been used yet, or the one that has been used longest time ago.

**Comment by FSSRepo** [comment]
> Can't we release the slot after every request?

You can try it, deleting `slot.params.cache_prompt ? SLEEPING : ` of:

```c++
            // release the slot
            if (slot.state == PROCESSING && slot.command == RELEASE)
            {
                slot.state = slot.params.cache_prompt ? SLEEPING : IDLE;
```

To this:

```c++
            // release the slot
            if (slot.state == PROCESSING && slot.command == RELEASE)
            {
                slot.state = IDLE;
```

This works with one client, but when a new client (doesn't have an slot it assigned) want to request a `completion` could use the released slot of the first client.

**Comment by ggerganov** [comment]
Ok, I think it works now.

I think the server actually worked out pretty good! Good job @FSSRepo, @monatis and all!
There are still some things to improve and the logic is a bit too complicated in some places, but we can gradually improve this.

Will merge this soon if no issues are found

File: examples/server/server.cpp
Comment (cebtenzzre) [inline examples/server/server.cpp:346]: Since this enum has only two values now, couldn't this be `bool idle = true` or `bool processing = false`?

**Comment by ggerganov** [comment]
Parallel `llava` is not working anymore - first client is OK, but starting a second one produces nonsense:

image

File: examples/server/server.cpp
Comment (cebtenzzre) [inline examples/server/server.cpp:440]: Can be simplified to `state == PROCESSING || command == LOAD_PROMPT` since there are only two possible states.

**Comment by FSSRepo** [comment]
> Parallel `llava` is not working anymore - first client is OK, but starting a second one produces nonsense:

I think that llava doen't work with parallel decoding

File: examples/server/public/index.html
Comment (monatis) [inline examples/server/public/index.html:455]: image slot id is statically set to 10 here. I think that might the reason for the parallel LLaVA issue.

Comment (ggerganov) [reply examples/server/public/index.html:455]: Ok, let's fix this later then

Comment (FSSRepo) [reply examples/server/public/index.html:455]: The images are loaded for each slot, so it doesn't matter if it's 10, as it ultimately loads a different slot.

**Comment by FSSRepo** [comment]
Screenshot 2023-10-22 152844
Unexpected behaivor. I didn't know I could send multiple messages about the same image, the only problem is that it has to load every time I ask something.

**Comment by monatis** [comment]
Yes technically we can chat about an image on multiple turns, and system prompt should not be prepended other than the first message, but I think optimizing it to eliminate the need to load/preprocess/encode images by a caching mechanism may be another story, i.e., another PR.

**Comment by cebtenzzre** [comment]
Why did this PR remove --n-threads-batch from the server? It only existed for one day (added in a8bdd65525ae86dea905e9866ad369b53e30ac14, removed in 29c8cdd65d1c6ced44f9f0d4f4c2d03a215ebed5).

**Comment by ggerganov** [comment]
This is my mistake - I tried to rebase this PR at some point but the changes to `master` were all around the place, so I applied them manually. Seems like I missed this one

**Comment by ibehnam** [comment]
> Continuation of #3589
> 
> Commands to test:
> 
> ```shell
> # parallel decoding from multiple clients (`-np 4`) with continuous batching (`-cb`)
> ./server -m ${model} --ctx_size 2048 -t 8 -ngl 99 -np 4 -cb
> 
> # vision support, upload image in the browser and ask questions
> ./server -m ./models/llava-7b-v1.5/ggml-model-f16.gguf --mmproj ./models/llava-7b-v1.5/mmproj-model-f16.gguf --ctx_size 2048 -t 8 -ngl 99
> ```

I can't thank you enough! This is great. I have one question:

- In my tests, `parallel` treats different lines in the prompt as **separate** prompts and sends each one to a different slot. For example:

```
prompt = '''llama is an animal that inspired work on large language models.\nThis is why we see a lot of LLM packages with the word "llama" in them.'''
```
In this case, and assuming that `np` is at least `2`, `parallel` processes `llama is an animal that inspired work on large language models.` and `This is why we see a lot of LLM packages with the word "llama" in them.` separately.

In my test, `server` doesn't show the same behavior 🤔 
Can you please provide more info about how to send multiple queries to the `server` with parallel processing?

**Comment by ibehnam** [comment]
I should mention that using `np` actually slowed down inference on my M1 Pro chip:

- Using parallelism:
```
./server -m "llama-2-13b.Q8_0.gguf" --ctx_size 4096 -ngl 128 --mlock --threads 10 --numa -np 16 -cb
```

-> 50 requests sent, total time = 60.38 seconds

- Not using parallelism:
```
./server -m "llama-2-13b.Q8_0.gguf" --ctx_size 4096 -ngl 128 --mlock --threads 8 --no-mmap --numa
```

-> 50 requests sent sequentially, total time = 52.73 seconds

**Comment by ggerganov** [comment]
Optimal performance when using parallelism with quantum models currently requires some manual adjustments to some constants depending on the model size, both for Metal and CUDA. I wrote some thoughts on this here https://github.com/ggerganov/llama.cpp/pull/3524 and here https://github.com/ggerganov/llama.cpp/pull/3545#issuecomment-1758128941

You can try playing with this piece of code to get better parallel performance for your use-case:

https://github.com/ggerganov/llama.cpp/blob/e3932593d46c30145301a13097895f9376cba509/ggml-metal.m#L1032-L1058

But atm it's not very user-friendly and I haven't found an universal way to obtain the best performance automatically.

**Comment by ibehnam** [comment]
@ggerganov Thanks, I will have a look into those. Something that occasionally helped speed up inference on my end was using native Python ThreadPoolExecutor to call the server endpoint in parallel. Although, similar to what you said, it required fiddling with the number of workers—sometimes parallel Python code execution would actually slow down inference.

**Pull request merged**
